bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/discovery/HDPowerViewShadeDiscoveryService.java

```diff
@@ -96,12 +97,35 @@ private Runnable createScanner() {
                                 String id = Integer.toString(shadeData.id);
                                 ThingUID thingUID = new ThingUID(HDPowerViewBindingConstants.THING_TYPE_SHADE,
                                         bridgeUID, id);
-                                DiscoveryResult result = DiscoveryResultBuilder.create(thingUID)
+                                DiscoveryResultBuilder builder = DiscoveryResultBuilder.create(thingUID)
                                         .withProperty(HDPowerViewShadeConfiguration.ID, id)
-                                        .withRepresentationProperty(HDPowerViewShadeConfiguration.ID)
-                                        .withLabel(shadeData.getName()).withBridge(bridgeUID).build();
+                                        .withRepresentationProperty(HDPowerViewShadeConfiguration.ID);
+
                                 logger.debug("Hub discovered shade '{}'", id);
-                                thingDiscovered(result);
+
+                                if (shadeData.type > 0) {
+                                    if (db.isTypeInDatabase(shadeData.type)) {
+                                        builder = builder.withProperty(HDPowerViewBindingConstants.PROPERTY_SHADE_TYPE,
+                                                db.getType(shadeData.type).toString());
+                                    } else {
+                                        db.logTypeNotInDatabase(shadeData.type);
+                                    }
+                                }
+
+                                Integer capabilities = shadeData.capabilities;
+                                int caps = (capabilities != null) ? capabilities.intValue() : -1;
+                                if (caps >= 0) {
+                                    if (db.isCapabilitiesInDatabase(caps)) {
+                                        builder = builder.withProperty(
+                                                HDPowerViewBindingConstants.PROPERTY_SHADE_CAPABILITIES,
+                                                db.getCapabilities(caps).toString());
```

Comment:
Two _O(n)_ lookups for the same capabilities right after each other. Consider dropping `isCapabilitiesInDatabase` and letting `getCapabilities` return some distinct value (or even null) if capabilities are not known to avoid two lookups.

Reply:
OK.